typing/typedtree.ml

```diff
@@ -627,41 +627,40 @@ let map_pattern_desc f d =
 
 (* List the identifiers bound by a pattern or a let *)
 
-let idents = ref([]: (Ident.t * string loc * Types.type_expr) list)
-
-let rec bound_idents pat =
+let rec iter_bound_idents f pat =
   match pat.pat_desc with
-  | Tpat_var (id,s) -> idents := (id,s,pat.pat_type) :: !idents
+  | Tpat_var (id,s) -> f (id,s,pat.pat_type)
   | Tpat_alias(p, id, s) ->
-      bound_idents p; idents := (id,s,pat.pat_type) :: !idents
+      iter_bound_idents f p; f (id,s,pat.pat_type)
   | Tpat_or(p1, _, _) ->
-      (* Invariant : both arguments binds the same variables *)
-      bound_idents p1
-  | d -> iter_pattern_desc bound_idents d
-
-let pat_bound_idents_full pat =
-  idents := [];
-  bound_idents pat;
-  let res = !idents in
-  idents := [];
-  res
+      (* Invariant : both arguments bind the same variables *)
+      iter_bound_idents f p1
+  | d -> iter_pattern_desc (iter_bound_idents f) d
 
-let pat_bound_idents pat =
-  List.map (fun (id,_,_) -> id) (pat_bound_idents_full pat)
+let rev_pat_bound_idents_full pat =
+  let idents_full = ref [] in
+  let add id_full = idents_full := id_full :: !idents_full in
+  iter_bound_idents add pat;
+  !idents_full
 
-let rev_let_bound_idents_with_loc bindings =
-  idents := [];
-  List.iter (fun vb -> bound_idents vb.vb_pat) bindings;
-  let res = !idents in idents := []; res
+let rev_only_idents idents_full =
+  List.rev_map (fun (id,_,_) -> id) idents_full
 
-let let_bound_idents_with_loc pat_expr_list =
-  List.rev(rev_let_bound_idents_with_loc pat_expr_list)
+let pat_bound_idents_full pat =
+  List.rev (rev_pat_bound_idents_full pat)
+let pat_bound_idents pat =
+  rev_only_idents (rev_pat_bound_idents_full pat)
```

Comment:
This two functions now return the list in the reverse order as before. I have checked all their callers (in the compiler, and in merlin), and I believe none of them care about the order.

But since we don't care about the order, why pay for the reversal? Perhaps you could rename the functions to `rev_pat_b...`? (Or not rename them at all?)

Reply:
(We only pay for a reversal in the `_full` case; in the non-full case it's cheaper to produce the identifiers in the right order, unless we were to write a reverse-iterator to do it in just one pass.)

I think that we should always return result in the natural order, it makes the API simpler and safer -- principle of least surprise -- especially in cases like here where performance does not matter.